File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.java
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int firstLength = 0;
        int secondLength = 0;
        ListNode headADummy = headA;
        ListNode headBDummy = headB;

        while (headADummy != null) {
            headADummy = headADummy.next;
            firstLength++;
        }
        while (headBDummy != null) {
            headBDummy = headBDummy.next;
            secondLength++;
        }

        if (firstLength > secondLength) {
            for (int i = 0; i < firstLength - secondLength; i++) {
                headA = headA.next;
            }
        } else {
            for (int i = 0; i < secondLength - firstLength; i++) {
                headB = headB.next;
            }
        }

        while (headA != null && headB != null) {
            if (headA == headB) {
                return headA;
            }
            headA = headA.next;
            headB = headB.next;
        }

        return null;
    }
}

File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* getIntersectionNode(ListNode* headA, ListNode* headB) {
        int firstSize = 0, secondSize = 0;
        ListNode* dummy1 = headA;
        ListNode* dummy2 = headB;
        while (dummy1) {
            firstSize++;
            dummy1 = dummy1->next;
        }
        while (dummy2) {
            secondSize++;
            dummy2 = dummy2->next;
        }
        if (firstSize > secondSize) {
            for (int i = 0; i < firstSize - secondSize; i++) {
                headA = headA->next;
            }
        } else if (firstSize < secondSize) {
            for (int i = 0; i < secondSize - firstSize; i++) {
                headB = headB->next;
            }
        }
        while (true) {
            if (headA == headB) {
                return headA;
            }
            headA = headA->next;
            headB = headB->next;
        }
        return headA;
    }
};